lib/semantics/check-do.cc

```diff
@@ -647,20 +643,124 @@ void DoChecker::Leave(const parser::DoConstruct &x) {
   doContext.Check(x);
 }
 
-// C1134
-void DoChecker::Enter(const parser::CycleStmt &) {
-  if (!context_.InsideDoConstruct()) {
-    context_.Say(
-        *context_.location(), "CYCLE must be within a DO construct"_err_en_US);
+// Return the (possibly null) name of the ConstructNode
+static const parser::Name *MaybeGetNodeName(const ConstructNode &construct) {
+  return std::visit(
+      [&](const auto &x) { return MaybeGetConstructName(*x); }, construct);
+}
+
+template<typename A> static parser::CharBlock GetConstructPosition(const A &a) {
+  return std::get<0>(a.t).source;
+}
+
+static parser::CharBlock GetNodePosition(const ConstructNode &construct) {
+  return std::visit(
+      [&](const auto &x) { return GetConstructPosition(*x); }, construct);
+}
+
+void DoChecker::SayBadLeave(StmtType stmtType, const char *enclosingStmtName,
+    const ConstructNode &construct) const {
+  context_
+      .Say("%s must not leave a %s statement"_err_en_US, EnumToString(stmtType),
+          enclosingStmtName)
+      .Attach(GetNodePosition(construct), "The construct that was left"_en_US);
+}
+
+static const parser::DoConstruct *MaybeGetDoConstruct(
+    const ConstructNode &construct) {
+  if (auto const *doNode{
+          std::get_if<const parser::DoConstruct *>(&construct)}) {
+    return *doNode;
+  }
+  return nullptr;
+}
+
+static bool ConstructIsDoConcurrent(const ConstructNode &construct) {
+  parser::DoConstruct const *doConstruct{MaybeGetDoConstruct(construct)};
+  return doConstruct && doConstruct->IsDoConcurrent();
+}
+
+// Check that CYCLE and EXIT statements do not cause flow of control to
+// leave DO CONCURRENT, CRITICAL, or CHANGE TEAM constructs.
+void DoChecker::CheckForBadLeave(
+    StmtType stmtType, const ConstructNode &construct) const {
+  std::visit(
+      common::visitors{
+          [&](const parser::DoConstruct *doConstructPtr) {
+            if (doConstructPtr->IsDoConcurrent()) {
+              // C1135 and C1167
+              SayBadLeave(stmtType, "DO CONCURRENT", construct);
+            }
+          },
+          [&](const parser::CriticalConstruct *) {
+            // C1135 and C1168
+            SayBadLeave(stmtType, "CRITICAL", construct);
+          },
+          [&](const parser::ChangeTeamConstruct *) {
+            // C1135 and C1168
+            SayBadLeave(stmtType, "CHANGE TEAM", construct);
+          },
+          [](const auto *) {},
+      },
+      construct);
+}
+
+static bool isExit(StmtType stmtType) { return stmtType == StmtType::EXIT; }
+
+static bool StmtMatchesConstruct(const parser::Name *stmtName,
+    StmtType stmtType, const parser::Name *constructName,
+    const ConstructNode &construct) {
+  bool inDoConstruct{MaybeGetDoConstruct(construct) != nullptr};
+  if (stmtName == nullptr) {
+    return inDoConstruct;  // Unlabeled statements match all DO constructs
+  } else if (constructName && constructName->source == stmtName->source) {
+    return isExit(stmtType) || inDoConstruct;
```

Comment:
It seems to me that `stmtType == StmtType::EXIT` is at least as clear as `isExit(stmtType)`.

If you keep `isExit` it should be `IsExit`.

Reply:
Thanks, Tim.  I'll remove and replace the function.